### ISSUE TITLE
Markdown Support

### ISSUE THREAD

**Issue by paulcalabro** [mrzapp/nextcloud-cookbook#47]
I really like this app!  Is it possible to add support for Markdown? I'd like to format the recipe a bit as well as reference multiple image files within the recipe itself.

**Comment by mrzapp** [comment]
I don't think it would be of much benefit, since Markdown just generates HTML, and that would just be noise in the JSON format that we're following: http://uischema.org/Recipe.

However, we could still add support for images on a per-step basis, since the "recipeInstructions" field can take an [ItemList](http://schema.org/ItemList) of [ListItem](http://schema.org/ListItem), which is based on [Thing](https://schema.org/Thing), which has an "image" field.

**TL;DR**: No markdown, yes images.

**Comment by sir-indy** [comment]
I've been thinking about this too, I think markdown would be useful, personally I like to mark recipe ingredients in **bold** (because I have a habit of skipping over them and not adding them).

**Comment by paulcalabro** [comment]
@mrzapp  That makes sense. Adding support for images in the steps would be a nice improvement!

**Comment by christianlupus** [comment]
> I don't think it would be of much benefit, since Markdown just generates HTML, and that would just be noise in the JSON format that we're following: http://uischema.org/Recipe.

@mrzapp: Just out of cuiosity: Whouldn't it be benefical to store the Markdown in the DB/JSON instead of the parsed HTML and do the Markdown->HTML conversion on the fly or in a separate cache? Then the JSON is comparably clean.

**Comment by mrzapp** [comment]
@christianlupus I see your point, that might work. That also makes markdown optional, which is nice. In most cases though, the recipes are fetched from a remote resource presenting HTML, which is stripped, so markdown is only really supported when entered manually.

I think we can just render the markdown client side, it's super quick.

**Comment by christianlupus** [comment]
I do not know how the stripping is done exactly. However I assume this is done in the regexes in [this file](https://github.com/mrzapp/nextcloud-cookbook/blob/master/lib/Service/RecipeService.php). Maybe one could tweak them to keep some basic formatting (like `<b>...</b>`) by simply replacing them with their markdown syntax before stripping.

**Comment by mrzapp** [comment]
@christianlupus yeah, it's [this](https://github.com/mrzapp/nextcloud-cookbook/blob/master/lib/Service/RecipeService.php#L164) line, which uses [this](https://github.com/mrzapp/nextcloud-cookbook/blob/master/lib/Service/RecipeService.php#L717) method.

If we want to preserve the formatting from fetched HTML, we might as well use [this](https://github.com/thephpleague/html-to-markdown) library to convert it to markdown and [this](https://github.com/erusev/parsedown) one to convert it back when presenting it.

You want to take a swing at this? Otherwise I'll get to it eventually.

**Comment by christianlupus** [comment]
@mrzapp I am a bit short in time at the moment as I just returned from holidays and I have some pressing things at hands. So I will unfortunately not find much time in the next few weeks to code much, sorry.

The libraries look really good. Put them as a submodule and much of the issue should be gone within a short period of time.

**Comment by mrzapp** [comment]
@christianlupus just to summarise the solution for this issue:

1. We store `recipeInstructions` as an array of arrays, making sure that at least the `text` field is present in each array.
2. When saving the recipe, we download images referenced by the `image` field (if any) to the recipe folder and delete any old ones.
  2.1 When saving the images, we should rename them to `step_$i.jpg` for ease of discovery.
3. In the UI, we include image upload/picker buttons for each step input.

How does that sound to you? Please feel free to challenge any of the points, they're not set in stone :)

**Comment by Bugsbane** [comment]
Not sure if this should be a separate feature request, but multiple images wouldn't only be useful for the steps, but for the main image itself. For example, before I make a recipe, I'll use the image from the website I got it from. After I've made a recipe, I'd love to add one or more photos of how it actually turned out. If a family member makes it, they might do the same.

**Comment by ShuiHuo** [comment]
Any progress?

**Comment by christianlupus** [comment]
Unfortunately no.

**Comment by sergeng** [comment]
this feature will make this app perfect!

**Comment by christianlupus** [comment]
We need a way to store the images, thus the dependency on #340.

**Comment by impeira** [comment]
The Markdown feature would be greatly appreciated! 😇

**Comment by gsantner** [comment]
After not using this for a long time and trying it again - is it really still not possible to add more than one (primary) image?
I don't see any attach/upload/multiple URL options, only one Image URL input.
Where can I find it, or really missing still?

Cookbook 0.11.7

Thanks everyone

**Comment by christianlupus** [comment]
It is still not implemented. Development has stalled due to personal reasons for a time. I hope to clean up in the future some technical debt then we can think about this.
